include/spblas/vendor/rocsparse/multiply.hpp

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <type_traits>
+
+#include <hip/hip_runtime.h>
+#include <rocsparse/rocsparse.h>
+
+#include <spblas/allocator.hpp>
+#include <spblas/detail/ranges.hpp>
+#include <spblas/detail/view_inspectors.hpp>
+
+#include "allocator.hpp"
+#include "types.hpp"
+
+namespace spblas {
+
+class spmv_state_t {
+public:
+  spmv_state_t() : spmv_state_t(std::make_shared<detail::rocm_allocator>()) {
+    rocsparse_create_handle(&handle_);
+  }
+
+  spmv_state_t(std::shared_ptr<const allocator> alloc)
+      : alloc_(alloc), buffer_size_(0) {
+    rocsparse_create_handle(&handle_);
+  }
+
+  ~spmv_state_t() {
+    alloc_->free(workspace_);
+    rocsparse_destroy_handle(handle_);
+  }
+
+  template <matrix A, vector B, vector C>
+    requires __detail::has_csr_base<A> &&
+             __detail::has_contiguous_range_base<B> &&
+             __ranges::contiguous_range<C>
+  void multiply(A&& a, B&& b, C&& c) {
+    auto a_base = __detail::get_ultimate_base(a);
+    auto b_base = __detail::get_ultimate_base(b);
+    using matrix_type = decltype(a_base);
+    using input_type = decltype(b_base);
+    using output_type = std::remove_reference_t<decltype(c)>;
+    using value_type = typename matrix_type::scalar_type;
+
+    auto alpha_optional = __detail::get_scaling_factor(a, b);
+    tensor_scalar_t<A> alpha = alpha_optional.value_or(1);
+
+    rocsparse_spmat_descr mat;
+    rocsparse_create_csr_descr(
```

Comment:
All rocsparse routines return a status. The status must be checked, and created objects here must be released if an error occurs.

Reply:
now all checked